### ISSUE TITLE
feature: patch header/metadata from JSON

### ISSUE THREAD

**Issue by bdon** [protomaps/go-pmtiles#63]
define a JSON schema like 
```
{
  'tile_type':0,
  'metadata':{
    'my_key':'my_value'
  }
}
```

to allow modification of pmtiles by writing a new file.

`pmtiles show --header-metadata-json` would also be useful to output this.

**Comment by dvd3v** [comment]
Is it always necessary to write to a new file? Given that PMTiles can become quite large, updating the metadata in place would be beneficial.

**Comment by bdon** [comment]
Yes, same as flatgeobuf, parquet or video files, there is no extra space left in the file.

**Comment by spatialillusions** [comment]
@bdon Since you sort of mentioned it in https://github.com/protomaps/basemaps/issues/160#issuecomment-1784156935, it would be really handy if I just could push all my extra data into the metadata json in a simple way.

**Comment by bdon** [comment]
the command would be something like `pmtiles metadata FILE.pmtiles --write metadata.json` where you provide whatever metadata.json you need. 

```
pmtiles metadata FILE.pmtiles > metadata.json
pmtiles metadata FILE.pmtiles --write metadata.json # this is a noop
```

It would be fast+atomic but require temporary disk space as large as the new file, since it writes a completely new file.

**Comment by bdon** [comment]
an open question is whether we ever want to expose the ability to modify the header (non-metadata parts) - I guess if you accidentally set the wrong bounds or tile type, or want to change the center position/zoom?

**Comment by bdon** [comment]
Step one is making `pmtiles show --metadata FILE` write just the raw JSON to stdout:

https://github.com/protomaps/go-pmtiles/pull/98

**Comment by AndrewJDR** [comment]
I have a question related to this -- Is it currently possible to add additional metadata to the output pmtiles file when running `pmtiles convert input.mbtiles  output.pmtiles`?

If this is not currently possible, I assume this would be a bit easier than an in-place method that works against existing pmtiles files?

**Comment by AndrewJDR** [comment]
(Incidentally, I'm not aware of any command line tool that can just patch the metadata of an existing mbtiles file, which would also solve this problem. If anyone is aware of something like this, please do let me know)

**Comment by bdon** [comment]
Yeah I would say use the `sqlite3` CLI to modify the MBTiles. there's no equivalent with the pmtiles CLI yet.

**Comment by bdon** [comment]
Needing this now, currently plan:

`pmtiles show foo.pmtiles --header-json > header.json`

Writes a JSON representation of the non-offset parts of the header to a file.

`pmtiles show foo.pmtiles --metadata > metadata.json`

Writes the JSON metadata to a file.

`pmtiles write-header old.pmtiles header.json new.pmtiles` (or maybe new.pmtiles first?)

Write the JSON representation of the header into a new archive. (maybe we need --inplace where it atomically deletes and renames?) All fields must be specified.

`pmtiles write-metadata old.pmtiles  metadata.json new.pmtiles` (or maybe new.pmtiles first?)

Write the JSON representation of the header into a new archive. (maybe --inplace) ?

**Comment by bdon** [comment]
This is implemented in v1.24. https://github.com/protomaps/go-pmtiles/releases/tag/v1.24.0

See CLI docs here: https://docs.protomaps.com/pmtiles/cli#edit
